design/TimelineFences.md

```diff
@@ -0,0 +1,95 @@
+# Timeline fences
+
+Having fences store a number internally and wait / signal numbers is what we call numerical fences.
+This is the design that D3D12 chose for `ID3D12Fence`.
+D3D12 allows waiting on a fence before it is signaled, but Vulkan disallows doing this on `VkSemaphore` because some OSes lack kernel primitive to wait-before-signal.
+This means that WebGPU will have to validate that any wait on a fence will be unblocked by a prior signal operation that has been enqueued.
+To simplify the validation of signal-before-wait, we can force signaled number to be strictly increasing.
+
+Thus each fence has two pieces of internal state:
+ - The signaled value, the latest value passed to a signal to the fence, which is also the greatest thanks to the monotonicity
+ - The completed value, the value corresponding to the lastest signal operation that has been executed.
+
+The fences will require additional restrictions and operations if WebGPU has multiple queues.
+These changes will be tagged with [multi-queue]
+
+# Creating fences
+
+To mirror native APIs, fences are created directly on the `WebGPUDevice`.
+
+```webidl
+dictionary WebGPUFenceDescriptor {
+    u64 initialValue = 0;
+
+    // [multi-queue]
+    WebGPUQueue signalQueue = null;
+};
+
+partial interface WebGPUFence {};
+
+partial interface WebGPUDevice {
+    WebGPUFence createFence(WebGPUFenceDescriptor descriptor);
+};
+```
+
+The fence is created with both internal values set to `initialValue` and [multi-queue] is restricted to be signaled on `signalQueue`.
+If `signalQueue` is set to `null`, it will act as if it was set to the "default queue" (pending multi-queue definition of what that is).
+
+# Signaling
+
+Signaling a fence is done like via this method:
+
+```webidl
+partial interface WebGPUQueue {
+    void signal(WebGPUFence fence, u64 signalValue);
+};
+```
+
+A Javascript exception is generated if:
+ - `value` is smaller or equal to the signaled value of the fence.
+ - [multi-queue] the fence must be signaled on the queue passed as the `signalQueue` of its descriptor.
+   This restriction is to make sure the signal operations are well-ordered, which would be more complicated if you could signal on multiple queues.
```

Comment:
Things would break down in cases like this. Let `F` be a timeline fence (current and signaled value at 0) and `Q1`, `Q2`, `Q3` be queues. The application issues the following:

 - `Q1.signal(F, 4)`
 - `Q2.signal(F, 5)`
 - `Q3.wait(F, 5)`

If `Q1` is scheduled before `Q2` then the fence value is 5 and `Q3` is unblocked. All is good.

If `Q2` is scheduled before `Q1` then the fence value is 4, and `Q3` is never unblocked, which isn't valid (all execution must finish in finite time, this is a good thing and also a requirement of Vulkan on Linux).

To avoid this case we would need to validate that the signal operation on `Q2` happens after the signal operation on `Q1`. This would require the application to put another fence between `Q1` and `Q2` and the implementation to walk the signaling graph to do the validation. This is complicated and I don't expect signaling on multiple queues to be a common use case at all.

Single-queue signaling helps because the ordering of the signal operations is obvious.

Reply:
Separately, if we ever add Signaling from the CPU to the API, then your example would become a problem.  Since CPU Signaling is done on the fence itself, as opposed to the queue, we can't enforce the "that has been enqueued" part of "any wait on a fence will be unblocked by a prior signal operation that has been enqueued" earlier in the document.

Signaling on the CPU allows us to do producer/consumer scenarios that involve long running GPU operations who's final step depends on a CPU upload.  Similarly, multi-queue GPU operations may produce textures that are readback by the CPU in a round robin fashion.  Both of these scenarios can still be done, but we would lose CPU/GPU parallelism.  

I do not think this is a big deal for MVP but something to keep in mind for the future.

Reply:
> Out of curiosity, what happens in Vulkan on Linux if you have multiple queues waiting for something that never gets signaled? Does the OS hang or crash or something else?

No idea :)

> Separately, if we ever add Signaling from the CPU to the API, then your example would become a problem. Since CPU Signaling is done on the fence itself, as opposed to the queue, we can't enforce the "that has been enqueued" part of "any wait on a fence will be unblocked by a prior signal operation that has been enqueued" earlier in the document.

This has the same issue that it isn't supported on Vulkan and isn't possible on Linux. I agree that it is an interesting use-case so if/when Vulkan has that capability we can think about it again. Also a fence on the CPU would probably be only signalable on the CPU.